Bulut/14 clean underlay

bbulgu opened the pull request.

yhassanzadeh13 commented on pom.xml:
```diff
@@ -51,12 +51,23 @@
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <version>3.9.0</version>
```
test comment

yhassanzadeh13 replied:
applied https://github.com/yhassanzadeh13/lightchain-container/pull/15/commits/80964f8ed8aab3387addd542ec345c9a9a415252

bbulgu commented:
- Added underlay class, which is used to abstract away the RMI primitives used in the skipGraph and lightchain nodes.
- Added tests for underlay.
- Created JavaDocs for changes.
- Reformatted code to comply with checkstyle linting. Used Google Checks.

yhassanzadeh13 commented on src/main/java/underlay/Underlay.java:
```diff
@@ -0,0 +1,168 @@
+public class Underlay {
+
+  /** The underlying RMIInterface instance. Is used for skipnode calls */
+  private RMIInterface targetRMI;
+
+  /** The underlying LightChainRMIInterface instance. Is used for lightchainnode class */
+  private LightChainRMIInterface targetLightChainRMI;
+
+  private final Logger logger = Logger.getLogger("");
```
These all should be set as part of the `Underlay`'s constructor. As a general best practice, we should not have separate setters for member fields that we do not expect to change and be updated during the runtime. Please also remove the setters.

bbulgu replied:
These (targetRMI and targetLightChainRMI) are set each time sendMessage is called - they change based on address, so I'm not sure if what you've said applies here.

yhassanzadeh13 replied:
Right, thanks.

yhassanzadeh13 commented on src/main/java/blockchain/ContractCV.java:
```diff
@@ -28,8 +33,7 @@ public ContractCV(LightChainNode owner) throws RemoteException {
             NodeInfo ndowner = owner.searchByNumID(t.getOwner());
-            LightChainRMIInterface ownerRMI = owner.getLightChainRMI(ndowner.getAddress());
-            int token1 = ownerRMI.getToken();
+            int token1 = ((IntegerResponse) underlay.sendMessage(new GetTokenRequest(), ndowner.getAddress(), InterfaceTypes.LightChainInterface)).result;
```
Also, please avoid having nested method calls, and separate the calls instead as much as possible. This also makes debugging easier. Please refactor it across all changes made in this PR: 

```
IntegerResponse response = IntegerResposonseOf(underlay.sendMessage(new GetTokenRequest(), ndowner.getAddress()));

int token1 = response.result;
```

yhassanzadeh13 commented on src/test/java/underlay/UnderlayTest.java:
```diff
@@ -0,0 +1,90 @@
+        NodeInfo result = new NodeInfo("1.1.1.1", 12, "test");
+        when(targetRMI.getLeftNode(1,2)).thenReturn(result);
+        assertEquals(result,
+                ((NodeInfoResponse) underlay.sendMessage(new GetLeftNodeRequest(1,2), "test")).responseResult);
```
A gentle reminder that nested calls should also be separated in tests.

yhassanzadeh13 commented on src/test/java/underlay/UnderlayTest.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+
+    // lightchainnode tests
```
Please remove this comment, instead add a `javadoc` for each test function saying what does the test function evaluate. A good example of writing java docs for test classes is available [here](https://docs.oracle.com/javame/test-tools/jdts/dev_guide/SpecialComments.html).

yhassanzadeh13 commented on src/test/java/underlay/UnderlayTest.java:
```diff
@@ -0,0 +1,90 @@
+    public void GetLeftNodeTest() throws RemoteException, FileNotFoundException {
+        RMIInterface targetRMI = mock(RMIInterface.class);
+        underlay.setTargetRMI(targetRMI);          // method created just for testing purposes
+        NodeInfo result = new NodeInfo("1.1.1.1", 12, "test");
```
This sort of object creation during the test is called _test fixturing_. As a best practice, we should not have fixture creating detail during our test. Instead, we better to have a `NodeInfoFixture` in our `src/text/java/fixture` class that creates _randomly_ generated fixtures for our tests. The fixtures should be meaningful, i.e., we expect the first parameter of `NodeInfo` constructor to be a valid IP address. Hence we better to implement `NodeInfoFixture` such that it generates a valid but random IP address for the node, as well as its other info. 

Then we can have `NodeInfo result = NodeInfoFixture()`. 

Please apply this to all tests in this test class, e.g., `TransactionFixture`, `SingedBytesFixture`, etc.

yhassanzadeh13 commented on src/test/java/underlay/UnderlayTest.java:
```diff
@@ -0,0 +1,90 @@
+        Transaction testTransaction = new Transaction("", 2, "","1",2);
+        SignedBytes testResult = new SignedBytes("test results".getBytes());
+        when(targetRMI.PoV(testTransaction)).thenReturn(testResult);
+        SignedBytes signature = ((SignatureResponse) underlay.sendMessage(new PoVRequest(testTransaction), "not_a_real_ip", InterfaceTypes.LightChainInterface)).response;
```
Following my previous comment, here we can have `IPAddressFixture` instead of `not_a_real_ip`. 
```
public static String IPAddressFixture(){
         Random r = new Random();
         return r.nextInt(256) + "." + r.nextInt(256) + "." + r.nextInt(256) + "." + r.nextInt(256);
}
```

yhassanzadeh13 reviewed (changes requested):
Great implementation, added some comments to improve the code quality, please apply them all and let me know. Thanks!

bbulgu commented on src/main/java/blockchain/LightChainCV.java:
```diff
@@ -1,8 +1,16 @@
 import java.rmi.RemoteException;
 
 class LightChainCV extends CorrectnessVerifier {
+    Underlay underlay = new Underlay();
```
I've changed this so that Underlay is initiated in UnderlayTest, LightChainNode and SkipNode.

bbulgu commented on src/main/java/underlay/Underlay.java:
```diff
@@ -0,0 +1,168 @@
+   * @throws RemoteException
+   * @throws FileNotFoundException
+   */
+  public GenericResponse sendMessage(GenericRequest req, String targetAddress, InterfaceTypes type)
```
I've changed the code based on your recommendations but I was wondering if the following would be even better:
Instead of the two methods you've outlined above calling the sendMessage function again, the sendMessage function with three parameters could instead call the methods above, ie the two parameter versions.

yhassanzadeh13 replied:
Please implement as we discussed.

bbulgu replied:
https://github.com/yhassanzadeh13/lightchain-container/pull/15/commits/4de632c13888c1d13a2b27665e2f81c20988f596, in this commit we use the sendmessage function with two parameters all over the code.

bbulgu replied:
Changed/fixed in https://github.com/yhassanzadeh13/lightchain-container/pull/15/commits/e97c50a210f563c29f346cb17288ab76631f533c. It turns out just one sendmessage function with the genericrequest parameter is enough.

yhassanzadeh13 commented on src/main/java/underlay/Underlay.java:
```diff
@@ -0,0 +1,168 @@
+ * The underlay class is used to abstract away the RMI primitives used in the skipGraph and
+ * lightchain nodes.
+ */
+public class Underlay {
```
Following part has not been applied: 

> In this way, we can keep the instance of underlay at node's components as Underlay abstract type, which is instantiated by RMIUnderlay, i.e., in the main class we will be passing Underlay underlay = new RMIUnderlay(....) to the Lightchain node.

We should not have patterns [like this](https://github.com/yhassanzadeh13/lightchain-container/blob/8522633464f6919e4091408ff7421644eea24e49/src/main/java/skipGraph/SkipNode.java#L83-L84). ⚠️ Indeed if check the usage, [this constructor](https://github.com/yhassanzadeh13/lightchain-container/blob/8522633464f6919e4091408ff7421644eea24e49/src/main/java/skipGraph/SkipNode.java#L56-L90) where we initialized an instance of `SkipNode` is _never used_. Instead the [other constructor](https://github.com/yhassanzadeh13/lightchain-container/blob/8522633464f6919e4091408ff7421644eea24e49/src/main/java/skipGraph/SkipNode.java#L100-L118) is utilized by derived child class. Hence:

🔥  we should make sure that each class has only a _single_ constructor: please merge-in duplicate constructors if needed and completely remove constructors that never used.

⚠️ we should have `Underlay` being passed as an interface type to the _constructor_ of the`SkipNode` and any other class that is needed. As a cardinal rule within all packages laying in the `src/main/java`, we should have an **overall single instance** of the `Underlay` initialized ([where we originally make an instance of the LightChain node](https://github.com/yhassanzadeh13/lightchain-container/blob/1be70f7af644d21d4727f159dcbc0bd6b6468840/src/main/java/simulation/Simulation.java#L31-L36)) and then pass it down the pipeline. So an acceptable `SkipNode` would look like: 

```
SkipNode(int port, int maxLevels, String introducer, Underlay underlay) throws RemoteException
```

bbulgu replied:
In https://github.com/yhassanzadeh13/lightchain-container/pull/15/commits/4de632c13888c1d13a2b27665e2f81c20988f596, we now have only one instance of Underlay initialized and passed through. The SkipNode constructor calls the other constructor now.

yhassanzadeh13 commented on src/test/java/underlay/RMIUnderlayTest.java:
```diff
@@ -0,0 +1,117 @@
+   */
+}
+
+//    @Test
```
Two comments about this file: 
- We should not have any commented code merged with master. If we are not sure about a code snippet, we should completely remove it instead of commenting it out.
- Every new feature (e.g., `RMIUnderlay`) we add, should be supported with unit tests. Now, are we losing unit test support for `RMIUnderlay`?

bbulgu replied:
Implemented in https://github.com/yhassanzadeh13/lightchain-container/pull/15/commits/e97c50a210f563c29f346cb17288ab76631f533c

yhassanzadeh13 commented on src/main/java/underlay/responses/IntegerResponse.java:
```diff
@@ -0,0 +1,10 @@
+
+/** Represents a response which returns an integer. */
+public class IntegerResponse extends GenericResponse {
+  public int result;
```
Comment has not been applied on requests in `src/main/java/underlay/requests/skipgraph`, e.g., [`GetLeftNumIDRequest`](https://github.com/yhassanzadeh13/lightchain-container/blob/f65eb2aa7675a1b75050b6a30016c056f4391cdd/src/main/java/underlay/requests/skipgraph/GetLeftNumIDRequest.java#L6-L7)

bbulgu replied:
Implemented in https://github.com/yhassanzadeh13/lightchain-container/pull/15/commits/426a2ee416cf3afc68fc8dfe88e6e61a7eba4a56
